src/v/kafka/server/quota_manager.cc

```diff
@@ -158,12 +178,37 @@ throttle_delay quota_manager::record_tp_and_throttle(
 
     auto prev = it->second.delay;
     it->second.delay = delay_ms;
-
     throttle_delay res{};
     res.first_violation = prev.count() == 0;
     res.duration = it->second.delay;
     return res;
 }
+
+// record a new observation and return <previous delay, new delay>
+throttle_delay quota_manager::record_produce_tp_and_throttle(
+  std::optional<std::string_view> client_id,
+  uint64_t bytes,
+  clock::time_point now) {
+    auto it = maybe_add_and_retrieve_quota(client_id, now);
+
+    it->second.tp_produce_rate.record(bytes, now);
+
+    return throttle(client_id, now, request_type::produce);
+}
+
+void quota_manager::record_fetch_tp(
+  std::optional<std::string_view> client_id,
+  uint64_t bytes,
+  clock::time_point now) {
+    auto it = maybe_add_and_retrieve_quota(client_id, now);
+    it->second.tp_fetch_rate.record(bytes, now);
+}
+
+throttle_delay quota_manager::throttle_fetch_tp(
+  std::optional<std::string_view> client_id, clock::time_point now) {
+    return throttle(client_id, now, request_type::fetch);
```

Comment:
I think it would be a right thing to call 
```
maybe_add_and_retrieve_quota(client_id, now)->second.tp_fetch_rate.record(0, now);
```
before calling `throttle()`. This way, `rate_tracker::maybe_advance_current()` will get called and advance the windows to the current time. Otherwise the position of the measurement windows may be stale and will not reflect the actual `now`.

Instead of passing `0` as the first arg to `record()`, there may be a separate function in `rate_tracker` for that too.

Reply:
I have missed the fact that `rate_tracker::measure()` purges the old windows (that is exactly what I was proposing), so please disregard.